src/component/parser/json/converter/ProcessConverter.ts

```diff
@@ -87,6 +90,26 @@ export default class ProcessConverter {
     }
   }
 
+  private assignIncomingAndOutgoingIdsFromFlows(): void {
+    const fillShapeBpmnElementAttribute = (
+      shapeBpmnElementId: string,
+      shapeBpmnElementAttributeName: keyof Pick<ShapeBpmnElement, 'outgoingIds' | 'incomingIds'>,
+      valueToAdd: string,
+    ): void => {
+      const shapeBpmnElement =
+        this.convertedElements.findFlowNode(shapeBpmnElementId) ?? this.convertedElements.findLane(shapeBpmnElementId) ?? this.convertedElements.findPoolById(shapeBpmnElementId);
+      if (shapeBpmnElement && !shapeBpmnElement[shapeBpmnElementAttributeName].includes(valueToAdd)) {
+        shapeBpmnElement[shapeBpmnElementAttributeName].push(valueToAdd);
+      }
+    };
+
+    const flows = [...this.convertedElements.getMessageFlows(), ...this.convertedElements.getSequenceFlows(), ...this.convertedElements.getAssociationFlows()];
```

Comment:
**question:** I guess we didn't measure the impact of this new lookup on the parsing performance in particular on large diagram.
The lookup done with `convertedElements.findXXX` involves retrieving elements from a Map, so it should be fast.
But having real figures would be better.

We could modify the `parseBpmn` utils to make it display the processing time or create a dedicated new tool for this purpose.

Reply:
As we decided with @tbouffard, we will manage the perfomance tests in another issue.

Reply:
After discussions with @csouchet, we decided to create a dedicated issue for the "parser performance": #2665
We will quickly handle it.
We have never done specific parsing performance test, so we have to setup the whole infrastructure. Performance tests involved the whole diagram loading and/or navigation.